ImageKit/Util/UploadApi.swift

```diff
@@ -8,67 +8,154 @@
+                let task = urlSession.dataTask(with: request) { data, response, error in
+                    if let error = error {
+                        completion(Result.failure(error))
+                        return
+                    }
+                    let response = response as! HTTPURLResponse
+                    let status = response.statusCode
+                    if (200...299).contains(status) {
+                        completion(Result.success((response, try? IKJSONDecoder().decode(UploadAPIResponse.self, from: data!))))
+                    } else if (400..<500).contains(status) || retryCount == uploadPolicy.maxErrorRetries {
+                        let uploadApiError = try? IKJSONDecoder().decode(UploadAPIError.self, from: data!)
+                        completion(Result.failure(uploadApiError!))
+                    } else {
+                        upload(
+                            file: file,
+                            token: token,
+                            fileName: fileName,
+                            useUniqueFileName: useUniqueFileName,
+                            tags: tags,
+                            folder: folder,
+                            isPrivateFile: isPrivateFile,
+                            progressClosure: progressClosure,
+                            urlConfiguration: urlConfiguration,
+                            uploadPolicy: uploadPolicy,
+                            completion: completion,
+                            retryCount: retryCount + 1
+                        )
+                    }
                 }
-                let response = response as! HTTPURLResponse
-                let status = response.statusCode
-                guard (200...299).contains(status) else {
-                    let uploadApiError = try? IKJSONDecoder().decode(UploadAPIError.self, from: data!)
-                    completion(Result.failure(uploadApiError!))
-                    return
+                task.resume()
+            } catch let error {
+                if retryCount == uploadPolicy.maxErrorRetries {
+                    completion(Result.failure(error))
+                } else {
+                    upload(
+                        file: file,
+                        token: token,
+                        fileName: fileName,
+                        useUniqueFileName: useUniqueFileName,
+                        tags: tags,
+                        folder: folder,
+                        isPrivateFile: isPrivateFile,
+                        progressClosure: progressClosure,
+                        urlConfiguration: urlConfiguration,
+                        uploadPolicy: uploadPolicy,
+                        completion: completion,
+                        retryCount: retryCount + 1
+                    )
                 }
-                completion(Result.success((response, try? IKJSONDecoder().decode(UploadAPIResponse.self, from: data!))))
             }
```

Comment:
Should we add a log line here that this was the error, retrying.

Also, we should have a handling here that if the error is retryable or not, if the response code is 403, that means the token is invalid, no point sending the same request again

Reply:
@ahnv this block is for performing the retries in case of errors like timeouts, network errors etc. Handling of errors from HTTP responses is being handled from line 108, where the check for HTTP 4xx responses is present.
I'll add the log statements for error here and everywhere else in this function.